### ISSUE TITLE
Table: Add ability to customize header and sorter

### ISSUE THREAD

**Issue by TuxujPes** [ant-design/ant-design#11246]
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
It seems like right now there is [no possibility to customize table's header](https://github.com/ant-design/ant-design/blob/master/components/table/Table.tsx#L760) clickables such as sorter and filter, as well as their positions.

E.g. A lot of users are used to sorters with toggle logic tied to header title and [sort icon](https://fontawesome.com/icons/sort-amount-up?style=solid) being changed upon that action. There is already an [open issue](https://github.com/ant-design/ant-design/issues/9911) for this particular example.

But there, probably, could be more and ideally a solution would be flexible enough to cover other cases.
Worth covering situations for when a specific column can be changed and not only on a table level.

With that being told, table should be able to provide ways to render custom header as well as to fire `sorter` callback.
Adding such API will give developers to build their tables upon antd more easily.

### What does the proposed API look like?
Below is a raw schematic API that supposedly explains the idea

```
static defaultProps = {
  ColumnHeader: (...args) => (
    <React.Fragment>
       {this.props.renderColumnTitle(...args)}
       {this.props.renderColumnSorter(...args)}
       {this.props.renderColumnFilter(...args)}
   </React.Fragment>
 )
}

...
// https://github.com/ant-design/ant-design/blob/bf5b6ae1f3b5da4629a7d10a2d1764eaa3fbbb04/components/table/Table.tsx#L760
column.title = (
  <span key={key}>
    {this.props.ColumnHeader(column, this.props, this, ...)}
  </span>
);



...
<Table
   renderColumnSorter: (col, props, ctx) => {
     const icon = col.sorter.order === 'ascend' ? 'up' : 'down';
     const direction = col.sorter.order === 'ascend' ? 'descend' : 'ascend;

      return <Button icon={icon} onClick={ctx.toggleSortOrder(direction) />
   }
```

**Comment by TuxujPes** [comment]
@afc163 
`column.title` can be either string or node.
But can it be function or class with access to data and internal state?

**Comment by afc163** [comment]
I see, we could support passing function to `title`:

```jsx
columns={[{
  title: (title, filters, sortOrder) => <div>{title}</div>,
}]}
```

**Comment by TuxujPes** [comment]
The idea is to extend current logic and functionality while `column.title` as right now gives the ability to change it completely.

**Comment by rahul-nath** [comment]
It would be great to be able to render an `Icon` as the Column title (or part of the column title).

**Comment by xiety** [comment]
sortOrder is passed the same for all columns, and there is no sortColumn to check with. Is it normal?

**Comment by fenghuizhang** [comment]
> I see, we could support passing function to `title`:
> 
> ```js-jsx
> columns={[{
>   title: (filters, sortOrder) => <div>Title</div>,
> }]}
> ```

No backward compatibility after version 3.10.0, before version 3.10.0 columns's title can be set to like 
`<span>xxx<span>*</span></span>`  before version 3.10.0,   but now its throw an error.
The reason is because if the title is not a function, just an React Element & has childNode.

**Comment by yoyo837** [comment]
@fongfai 
![image](https://user-images.githubusercontent.com/6134547/64547636-8a783b80-d35f-11e9-82cd-d46c40f7f1ff.png)

**Comment by swillis12** [comment]
As @xiety said, there is no way to identify the column that is currently sorted in the `title` render function. We get an object that has either "ascend", "descend", but no information to identify which actual column is sorted.

**Comment by RixBai** [comment]
@afc163 What's the solution to customize sort icon with sorting status?

**Comment by RixBai** [comment]
Oh.. Sorry @afc163  I was able to get the sort order with _title_ property.
Thank you
![image](https://user-images.githubusercontent.com/43899212/76581576-4a2e7d00-64aa-11ea-9981-495975a9f87c.png)

**Comment by vavdav** [comment]
@RixBai how do you hide default sort icons in the header?

**Comment by Abhikumar98** [comment]
> Oh.. Sorry @afc163 I was able to get the sort order on **_title_** property.
> Thank you
> ![image](https://user-images.githubusercontent.com/43899212/76581576-4a2e7d00-64aa-11ea-9981-495975a9f87c.png)
> ![image](https://user-images.githubusercontent.com/43899212/76582983-51578a00-64ae-11ea-9dfe-bda1008fdef7.png)

@RixBai How are you able to override the default sort icons ? Can you post snippet please ?

**Comment by hyungjunk** [comment]
Controlling visibility via CSS was my best shot to hide the sorting buttons.

```
span.ant-table-column-sorter-inner {
    visibility: hidden;
}
```

**Comment by hyungjunk** [comment]
@vavdav @Abhikumar98 

TL;DR;

Adding `sortDirections: ["YOUR_ORDER_NAME1", "YOUR_ORDER_NAME2"]` options to column makes the sorting arrow button hidden.

What I did is

```
const Name = {
  title: "Name",
  dataIndex: "name",
  sorter: true,
  sortDirections: ["ASC", "DESC"],
};
```
If `YOUR_ORDER_NAME `remains the same as original as 'ascend' and 'descend', the button is still there.

Other than those two names, the button will be gone.

And be careful that the value in sort field won't be `ascend `and `descend `anymore but your custom name.

```
<Table
...
onChange={onChange} />

const onChange = (pagination, filters, sorter, extra) => {
   console.log(sorter);
   // value in order field is not ascend and descend anymore but how you described.
}
```

**Comment by danielo515** [comment]
I would love to receive the column selector and the applied filters on the title function. The information provided right now is quite poor

**Comment by sagar-siwakoti** [comment]
hi ,can i also get solution on how to change icon of sorter with custom icon

**Comment by siowzhenhan** [comment]
> What I did is
> 
> ```
> const Name = {
>   title: "Name",
>   dataIndex: "name",
>   sorter: true,
>   sortDirections: ["ASC", "DESC"],

@hyungjunk I tried this, but now the column can only be sorted either in a _descending_ manner or _undefined_ manner, how did you add _ascending_ order with this method?

**Comment by souravdasdip** [comment]
<Tabs
        tabBarStyle={{ fontWeight: 'bold' }}
>
//content
</Tabs>
